Gradle 6.8 incorrectly swaps binary dependency to project dependency

<!--- 
Please use our bug report template to report problems with something that has never worked.  
Regressions reports are greatly appreciated during our RC phase and before a final release.
-->

When upgrading from Gradle 6.7.1 to 6.8 we found that our project started failing with
```
aurimas@aurimas:/ssd/ssd1/temp/testgradle68$ ./gradlew arch:core:core-common:jar
Watching the file system is not supported on this operating system.
Configuration on demand is an incubating feature.

FAILURE: Build failed with an exception.

* What went wrong:
Could not determine the dependencies of task ':arch:core:core-common:compileJava'.
> Could not resolve all task dependencies for configuration ':arch:core:core-common:compileClasspath'.
   > Could not resolve androidx.annotation:annotation:1.1.0.
     Required by:
         project :arch:core:core-common
      > No matching configuration of project :annotation:annotation was found. The consumer was configured to find an API of a library compatible with Java 7, preferably in the form of class files, and its dependencies declared externally but:
          - None of the consumable configurations have attributes.
```

There were no other changes to our build outside of Gradle upgrade.

This project uses the following dependency
```
dependencies {
    api("androidx.annotation:annotation:1.1.0")
}
```

### Expected Behavior
Gradle uses binary dependency for this project


### Current Behavior
Gradle swaps this project dependency to be `project :annotation:annotation`

### Context
There seems to be several factors at play that are required to reproduce this failure
- Use of `includeBuild`
- Rename of project in `settings.gradle`
- Use of `org.gradle.configureondemand=true`

It works just fine in Gradle 6.7.1 and it fails with 6.8

### Steps to Reproduce 
Attaching a sample project
[testgradle68.zip](https://github.com/gradle/gradle/files/5810783/testgradle68.zip)

To repro simply run:
```
./gradlew arch:core:core-common:jar
```

If you downgrade to 6.7.1 it will work again.

### Your Environment
Unable to share full details due to corp policy.

--- a/subprojects/composite-builds/src/integTest/groovy/org/gradle/integtests/composite/CompositeBuildDependencyGraphIntegrationTest.groovy
+++ b/subprojects/composite-builds/src/integTest/groovy/org/gradle/integtests/composite/CompositeBuildDependencyGraphIntegrationTest.groovy
@@ -777,6 +777,61 @@ Required by:
         failure.assertHasCause("Could not find test-1.2.jar (org.test:test:1.2).")
     }
 
+    @Unroll
+    def "substitutes external dependency for a subproject of the root build - #rootIsIncluded"() {
+        given:
+        def empty = file('empty').tap {
+            it.mkdir()
+        }
+        mavenRepo.module("org.test", "subproject1", "2.0").publish()
+        buildA.buildFile << """
+            subprojects {
+                apply plugin: 'java-library'
+                group = 'org.test'
+                version = '1.0'
+            }
+            dependencies {
+                implementation "org.test:subproject1:2.0"
+            }
+        """
+        includedBuilds = [empty]
+        buildA.settingsFile << """
+            include('subproject1')
+            $includeRootStatement
+        """
+
+        when:
+        checkDependencies()
+
+        then:
+        def expectSubstitution = rootIsIncluded.contains("yes")
+        checkGraph {
+            if (expectSubstitution) {
+                edge("org.test:subproject1:2.0", "project :subproject1", "org.test:subproject1:1.0") {
+                    configuration = "runtimeElements"
+                    compositeSubstitute()
+                }
+            } else {
+                module("org.test:subproject1:2.0")
+            }
+        }
+
+        and:
+        if (expectSubstitution) {
+            executed(":subproject1:jar")
+        } else {
+            notExecuted(":subproject1:jar")
+        }
+
+        where:
+        rootIsIncluded          | includeRootStatement
+        'no'                    | ""
+        'yes'                   | "includeBuild('.')"
+        // If substitutions for the root would be controllable (e.g. by the first include statement encountered) this would enable you to not have the ':subproject1' substitution
+        // This documents the current behavior. It is unclear if this is a useful functionality to have.
+        'yes with substitution' | "includeBuild('.') { dependencySubstitution { substitute module('org.test:subproject2') with project(':subproject2') } }"
+    }
+
     private void withArgs(List<String> args) {
         buildArgs = args as List
     }

--- a/subprojects/composite-builds/src/integTest/groovy/org/gradle/integtests/composite/CompositeBuildIncludeCycleIntegrationTest.groovy
+++ b/subprojects/composite-builds/src/integTest/groovy/org/gradle/integtests/composite/CompositeBuildIncludeCycleIntegrationTest.groovy
@@ -173,6 +173,7 @@ class CompositeBuildIncludeCycleIntegrationTest extends AbstractCompositeBuildIn
         """
         buildA.settingsFile << """
             rootProject.name = 'theNameOfBuildA'
+            includeBuild '../rootBuild'
         """
 
         when:
@@ -195,6 +196,7 @@ class CompositeBuildIncludeCycleIntegrationTest extends AbstractCompositeBuildIn
                 }
             """
             settingsFile << """
+                includeBuild '.'
                 includeBuild '../buildA'
                 includeBuild '../buildB'
             """
@@ -210,6 +212,9 @@ class CompositeBuildIncludeCycleIntegrationTest extends AbstractCompositeBuildIn
         buildB.buildFile << """
             apply plugin: 'java-library'
             dependencies { api 'org.test:x3' }
+        """
+        buildB.settingsFile << """
+            includeBuild '../rootBuild'
         """
 
         when:

--- a/subprojects/composite-builds/src/integTest/groovy/org/gradle/integtests/composite/CompositeBuildRootProjectIntegrationTest.groovy
+++ b/subprojects/composite-builds/src/integTest/groovy/org/gradle/integtests/composite/CompositeBuildRootProjectIntegrationTest.groovy
@@ -20,7 +20,7 @@ import org.gradle.integtests.fixtures.build.BuildTestFile
 
 class CompositeBuildRootProjectIntegrationTest extends AbstractCompositeBuildIntegrationTest {
 
-    def "root of a composite build can refer to own subprojects by GA coordinates"() {
+    def "root of a composite build cannot refer to own subprojects by GA coordinates by default"() {
         given:
         def buildB = multiProjectBuild("buildB", ['c1', 'c2', 'c3']) {
             buildFile << """
@@ -34,7 +34,32 @@ class CompositeBuildRootProjectIntegrationTest extends AbstractCompositeBuildInt
         dependency(new BuildTestFile(buildB.file('c3'), 'c3'), "org.test:buildB") // dependency to root
 
         buildB.settingsFile << """
-            includeBuild('${buildA.toURI()}')
+            includeBuild('${buildA.toURI()}') // include another build to become a composite
+        """
+
+        when:
+        fails(buildB, "c3:jar")
+
+        then:
+        failure.assertHasCause("Cannot resolve external dependency org.test:buildB because no repositories are defined.")
+    }
+
+    def "root of a composite build can refer to own subprojects by GA coordinates when including itself"() {
+        given:
+        def buildB = multiProjectBuild("buildB", ['c1', 'c2', 'c3']) {
+            buildFile << """
+            allprojects {
+                apply plugin: 'java-library'
+            }
+            """
+        }
+        dependency(buildB, "org.test:c1")
+        dependency(buildB, "org.test:c2:1.0")
+        dependency(new BuildTestFile(buildB.file('c3'), 'c3'), "org.test:buildB") // dependency to root
+
+        buildB.settingsFile << """
+            includeBuild('${buildA.toURI()}') // include another build to become a composite
+            includeBuild('.')
         """
 
         when:
@@ -59,6 +84,9 @@ class CompositeBuildRootProjectIntegrationTest extends AbstractCompositeBuildInt
            tasks.register("buildBJar") {
                 dependsOn(gradle.includedBuild("buildB").task(":jar"))
             }
+        """
+        buildA.settingsFile << """
+            includeBuild('.')
         """
 
         includeBuild(buildB)

--- a/subprojects/composite-builds/src/main/java/org/gradle/composite/internal/DefaultIncludedBuildRegistry.java
+++ b/subprojects/composite-builds/src/main/java/org/gradle/composite/internal/DefaultIncludedBuildRegistry.java
@@ -63,6 +63,7 @@ public class DefaultIncludedBuildRegistry implements BuildStateRegistry, Stoppab
     private final Map<File, IncludedBuildState> includedBuildsByRootDir = new LinkedHashMap<>();
     private final Map<Path, File> includedBuildDirectoriesByPath = new LinkedHashMap<>();
     private final Deque<IncludedBuildState> pendingIncludedBuilds = new ArrayDeque<>();
+    private boolean registerSubstitutionsForRootBuild = false;
 
     public DefaultIncludedBuildRegistry(BuildTreeState owner, IncludedBuildFactory includedBuildFactory, IncludedBuildDependencySubstitutionsBuilder dependencySubstitutionsBuilder, GradleLauncherFactory gradleLauncherFactory, ListenerManager listenerManager) {
         this.owner = owner;
@@ -162,7 +163,7 @@ public class DefaultIncludedBuildRegistry implements BuildStateRegistry, Stoppab
 
     @Override
     public void afterConfigureRootBuild() {
-        if (!includedBuildsByRootDir.isEmpty()) {
+        if (registerSubstitutionsForRootBuild) {
             dependencySubstitutionsBuilder.build((CompositeBuildParticipantBuildState) rootBuild);
         }
     }
@@ -207,6 +208,11 @@ public class DefaultIncludedBuildRegistry implements BuildStateRegistry, Stoppab
         // Attach the build only after it has been fully constructed.
         rootOfNestedBuildTree.attach();
         return rootOfNestedBuildTree;
+    }
+
+    @Override
+    public void registerSubstitutionsForRootBuild() {
+        registerSubstitutionsForRootBuild = true;
     }
 
     private void validateNameIsNotBuildSrc(String name, File dir) {

--- a/subprojects/core/src/main/java/org/gradle/internal/build/BuildStateRegistry.java
+++ b/subprojects/core/src/main/java/org/gradle/internal/build/BuildStateRegistry.java
@@ -102,4 +102,9 @@ public interface BuildStateRegistry {
      * Creates a new standalone nested build tree.
      */
     NestedRootBuild addNestedBuildTree(BuildDefinition buildDefinition, BuildState owner, @Nullable String buildName);
+
+    /**
+     * Register dependency substitutions for the root build itself. This way, the projects of the root build can be addressed by coordinates as the projects of all other builds.
+     */
+    void registerSubstitutionsForRootBuild();
 }

--- a/subprojects/core/src/main/java/org/gradle/internal/composite/ChildBuildRegisteringSettingsLoader.java
+++ b/subprojects/core/src/main/java/org/gradle/internal/composite/ChildBuildRegisteringSettingsLoader.java
@@ -61,6 +61,7 @@ public class ChildBuildRegisteringSettingsLoader implements SettingsLoader {
                     IncludedBuildState includedBuild = addIncludedBuild(includedBuildSpec, gradle);
                     children.add(includedBuild.getModel());
                 } else {
+                    buildRegistry.registerSubstitutionsForRootBuild();
                     children.add(new IncludedRootBuild((CompositeBuildParticipantBuildState) buildRegistry.getRootBuild()));
                 }
             }

--- a/subprojects/docs/src/docs/userguide/authoring-builds/software-product/composite_builds.adoc
+++ b/subprojects/docs/src/docs/userguide/authoring-builds/software-product/composite_builds.adoc
@@ -141,6 +141,9 @@ Source dependencies are configured, providing cross-build navigation and refacto
 By default, Gradle will configure each included build in order to determine the dependencies it can provide.
 The algorithm for doing this is very simple: Gradle will inspect the group and name for the projects in the included build, and substitute project dependencies for any external dependency matching `${project.group}:${project.name}`.
 
+NOTE: By default, substitutions are not registered for the _main_ build.
+To make the (sub)projects of the main build addressable by `${project.group}:${project.name}`, you can tell Gradle to treat the main build like an included build by self-including it: `includeBuild(".")`.
+
 There are cases when the default substitutions determined by Gradle are not sufficient, or they are not correct for a particular composite.
 For these cases it is possible to explicitly declare the substitutions for an included build.
 Take for example a single-project build 'anonymous-library', that produces a java utility library but does not declare a value for the group attribute:

